README.md
# dbjs-mongo
##  [mongodb](https://www.mongodb.org/) driver for [dbjs](https://github.com/medikoo/dbjs#dbjs)

## Tests [![Build Status](https://travis-ci.org/medikoo/dbjs-mongo.svg)](https://travis-ci.org/medikoo/dbjs-mongo)

	$ npm test


CHANGES
v0.0.16  --  2016.07.01
* Fix pass of id in searchComputed

v0.0.15  --  2016.03.18
* Update up to changes in dbjs-persistence

v0.0.14  --  2016.01.28
* Update up to changes in dbjs-persistence
  (implementation of global reduced storage)

v0.0.13  --  2016.01.22
* Update up to changes in dbjs-persistence
  (driver & storage split)
* Lazy collections resolution

v0.0.12  --  2015.12.30
* Update up to changes in dbjs-persistence

v0.0.11  --  2015.12.29
* Update up to changes in dbjs-persistence

v0.0.10  --  2015.12.01
* Update to use v0.0.12 of dbjs-persistence

v0.0.9  --  2015.12.01
* Update up to changes in dbjs-persistence

v0.0.8  --  2015.11.18
* Upgrade up to changes in dbjs-persistence
* Use separate collections for direct, computed and reduced data
* Store data in format friendly for direct mongodb processing

v0.0.7  --  2015.10.25
* Upgrade up to changes in dbjs-persistence

v0.0.6  --  2015.10.22
* Update dbjs-persistence dependency

v0.0.5  --  2015.10.19
* Improve emit of progress events in loadAll

v0.0.4  --  2015.10.17
* Ensure progress events on output promise

v0.0.3  --  2015.10.17
* Update up to changes in dbjs-persistence

v0.0.2  --  2015.10.15
* Implement export methods
* Ensure proper promises resolution
* Improve tests
* Update up to changes in dbjs-persistence

v0.0.1  --  2015.10.02
* Initial


test/index.js
'use strict';

var getTests          = require('dbjs-persistence/test/_common')
  , storageSplitTests = require('dbjs-persistence/test/_storage-split')

  , env, copyEnv, splitEnv, tests;

var genPostfix = function () {
	return (new Date()).toISOString().replace(/\./g, '-').replace(/:/g, '-');
};

try {
	env = require('../env');
} catch (e) {
	env = null;
}

if (env) {
	env.database = 'dbjs-mongo-test-' + genPostfix();
	copyEnv = Object.create(env);
	copyEnv.database = 'dbjs-mongo-test-copy-' + genPostfix();

	splitEnv = Object.create(env);
	splitEnv.database = 'dbjs-mongo-split-' + genPostfix();

	tests = getTests({ mongo: env }, { mongo: copyEnv });
}

module.exports = function (t, a, d) {
	if (!env) {
		console.error("No database configuration (env.json), unable to proceed with test");
		d();
		return;
	}
	return tests.apply(null, arguments)(function () {
		return storageSplitTests(t, { mongo: splitEnv }, a);
	}).done(function () { d(); }, d);
};


test/storage.js
// Tested in scope of ./driver.js

'use strict';

module.exports = Function.prototype;


index.js
'use strict';

var setPrototypeOf = require('es5-ext/object/set-prototype-of')
  , ensureObject   = require('es5-ext/object/valid-object')
  , ensureString   = require('es5-ext/object/validate-stringifiable-value')
  , d              = require('d')
  , deferred       = require('deferred')
  , Driver         = require('dbjs-persistence/driver')
  , Mongo          = require('mongodb')
  , Storage        = require('./storage')
  , ReducedStorage = require('./reduced-storage')

  , isIdent = RegExp.prototype.test.bind(/^[a-z][a-z0-9A-Z]*$/)

  , promisify = deferred.promisify
  , MongoCursor = Mongo.Cursor, MongoClient = Mongo.MongoClient
  , connect = promisify(MongoClient.connect);

Object.defineProperties(MongoCursor.prototype, {
	nextPromised: d(promisify(MongoCursor.prototype.next)),
	toArrayPromised: d(promisify(MongoCursor.prototype.toArray)),
	closePromised: d(promisify(MongoCursor.prototype.close))
});

var buildUrl = function (conf) {
	var url = 'mongodb://';
	if (conf.user && conf.password) url += conf.user + ':' + conf.password + '@';
	url += (conf.host != null) ? conf.host : 'localhost';
	url += ':';
	url += (conf.port != null) ? conf.port : '27017';
	return url + '/' + conf.database;
};
var MongoDriver = module.exports = Object.defineProperties(function (data) {
	if (!(this instanceof MongoDriver)) return new MongoDriver(data);
	ensureObject(data);
	ensureObject(data.mongo);
	ensureString(data.mongo.database);
	Driver.call(this, data);
	this.mongoDb = connect(buildUrl(data.mongo)).aside(null, this.emitError);
}, {
	storageClass: d(Storage),
	reducedStorageClass: d(ReducedStorage)
});
setPrototypeOf(MongoDriver, Driver);

MongoDriver.prototype = Object.create(Driver.prototype, {
	constructor: d(MongoDriver),

	__resolveAllStorages: d(function () {
		return this.mongoDb.invokeAsync('collections')(function (collections) {
			collections.forEach(function (collection) {
				var name = collection.collectionName;
				if (!isIdent(name)) return;
				this.getStorage(name);
			}, this);
		}.bind(this))(Function.prototype);
	}),
	__close: d(function () { return this.mongoDb.invokeAsync('close'); })
});


reduced-storage.js
'use strict';

var assign           = require('es5-ext/object/assign')
  , constant         = require('es5-ext/function/constant')
  , setPrototypeOf   = require('es5-ext/object/set-prototype-of')
  , serializeValue   = require('dbjs/_setup/serialize/value')
  , unserializeValue = require('dbjs/_setup/unserialize/value')
  , d                = require('d')
  , lazy             = require('d/lazy')
  , deferred         = require('deferred')
  , ReducedStorage   = require('dbjs-persistence/reduced-storage')

  , create = Object.create
  , getNull = constant(null)
  , isUnserializable = RegExp.prototype.test.bind(/[01234]/)
  , updateOpts = { upsert: true };

var MongoReducedStorage = module.exports = function (driver) {
	if (!(this instanceof MongoReducedStorage)) return new MongoReducedStorage(driver);
	ReducedStorage.call(this, driver);
};
setPrototypeOf(MongoReducedStorage, ReducedStorage);

MongoReducedStorage.prototype = Object.create(ReducedStorage.prototype, assign({
	constructor: d(MongoReducedStorage),

	// Any data
	__get: d(function (ns, path) { return this._get_(ns + (path ? ('/' + path) : '')); }),
	__store: d(function (ns, path, data) { return this._store_(ns, path, data); }),
	__getObject: d(function (ns, keyPaths) {
		return this.reducedDb.invokeAsync('find', { ns: ns })(function (cursor) {
			return cursor.toArrayPromised()(function (records) {
				var result = create(null);
				records.forEach(function (record) {
					result[record._id] = {
						value: record.unserialized ? serializeValue(record.value) : record.value,
						stamp: record.stamp
					};
				});
				return cursor.closePromised()(result);
			}.bind(this));
		}.bind(this));
	}),

	// Storage import/export
	__exportAll: d(function (destStorage) {
		var count = 0;
		var promise = deferred(
			this.reducedDb.invokeAsync('find')(function (cursor) {
				return cursor.toArrayPromised()(function (records) {
					return cursor.closePromised()(deferred.map(records, function (record) {
						if (!(++count % 1000)) promise.emit('progress');
						return destStorage._storeRaw(record.ns, record.path, {
							value: record.unserialized ? serializeValue(record.value) : record.value,
							stamp: record.stamp
						});
					}, this));
				}.bind(this));
			}.bind(this))
		);
		return promise;
	}),
	__clear: d(function () {
		return deferred(this.hasOwnProperty('reducedDb') && this.reducedDb.invokeAsync('drop'));
	}),
	__drop: d(function () { return this.__clear(); }),

	// Connection related
	__close: d(function () { return deferred(undefined); }),

	// Driver specific
	_get_: d(function (key) {
		return this.reducedDb.invokeAsync('find', { _id: key })(function (cursor) {
			return cursor.nextPromised()(function (record) {
				return cursor.closePromised()(record ? {
					value: record.unserialized ? serializeValue(record.value) : record.value,
					stamp: record.stamp
				} : getNull);
			});
		});
	}),
	_store_: d(function (ns, path, data) {
		var unserialized = Boolean(data.value && isUnserializable(data.value[0]));
		var record = {
			ns: ns,
			value: unserialized ? unserializeValue(data.value) : data.value,
			stamp: data.stamp
		};
		if (path) record.path = path;
		if (unserialized) record.unserialized = true;
		return this.reducedDb.invokeAsync('update', { _id: ns + (path ? ('/' + path) : '') },
			record, updateOpts);
	})
}, lazy({
	reducedDb: d(function () { return this.driver.mongoDb.invokeAsync('collection', '_reduced'); })
})));


storage.js
'use strict';

var assign              = require('es5-ext/object/assign')
  , constant            = require('es5-ext/function/constant')
  , setPrototypeOf      = require('es5-ext/object/set-prototype-of')
  , serializeValue      = require('dbjs/_setup/serialize/value')
  , unserializeValue    = require('dbjs/_setup/unserialize/value')
  , resolveKeyPath      = require('dbjs/_setup/utils/resolve-key-path')
  , d                   = require('d')
  , lazy                = require('d/lazy')
  , deferred            = require('deferred')
  , Storage             = require('dbjs-persistence/storage')
  , resolveValue        = require('dbjs-persistence/lib/resolve-direct-value')
  , filterComputedValue = require('dbjs-persistence/lib/filter-computed-value')

  , create = Object.create
  , getNull = constant(null)
  , isUnserializable = RegExp.prototype.test.bind(/[01234]/)
  , updateOpts = { upsert: true };

var MongoStorage = module.exports = function (driver, name/*, options*/) {
	if (!(this instanceof MongoStorage)) return new MongoStorage(driver, name, arguments[2]);
	Storage.call(this, driver, name, arguments[2]);
};
setPrototypeOf(MongoStorage, Storage);

MongoStorage.prototype = Object.create(Storage.prototype, assign({
	constructor: d(MongoStorage),

	// Any data
	__getRaw: d(function (cat, ns, path) {
		if (cat === 'reduced') return this._getReduced_(ns + (path ? ('/' + path) : ''));
		if (cat === 'computed') return this._getComputed_(path, ns);
		return this._getDirect_(ns, path);
	}),
	__storeRaw: d(function (cat, ns, path, data) {
		if (cat === 'reduced') return this._storeReduced_(ns, path, data);
		if (cat === 'computed') return this._storeComputed_(path, ns, data);
		return this._storeDirect_(ns, path, data);
	}),

	// Direct data
	__getObject: d(function (ownerId, objectPath, keyPaths) {
		var query;
		if (objectPath) {
			query = {
				_id: { $gte: ownerId + '/' + objectPath, $lt:  ownerId + '/' + objectPath + '/\uffff' }
			};
		} else {
			query = { ownerId: ownerId };
		}
		return this._loadDirect_(query, keyPaths && function (ownerId, path) {
			if (!path) return true;
			return keyPaths.has(resolveKeyPath(ownerId + '/' + path));
		});
	}),
	__getAllObjectIds: d(function () {
		return this.directDb.invokeAsync('find')(function (cursor) {
			return cursor.toArrayPromised()(function (records) {
				var data = create(null);
				records.forEach(function (record) {
					if (!record.keyPath) data[record.ownerId] = record;
				});
				return cursor.closePromised()(data);
			});
		}.bind(this));
	}),
	__getAll: d(function () { return this._loadDirect_(); }),

	// Reduced data
	__getReducedObject: d(function (ns, keyPaths) {
		return this.reducedDb.invokeAsync('find', { ns: ns })(function (cursor) {
			return cursor.toArrayPromised()(function (records) {
				var result = create(null);
				records.forEach(function (record) {
					result[record._id] = {
						value: record.unserialized ? serializeValue(record.value) : record.value,
						stamp: record.stamp
					};
				});
				return cursor.closePromised()(result);
			}.bind(this));
		}.bind(this));
	}),

	__search: d(function (keyPath, value, callback) {
		return this.directDb.invokeAsync('find')(function (cursor) {
			return cursor.toArrayPromised()(function (records) {
				records.some(function (record) {
					var recordValue, resolvedValue;
					if (keyPath !== undefined) {
						if (!keyPath) {
							if (record.keyPath) return;
						} else if (keyPath !== record.keyPath) {
							return;
						}
					}
					recordValue = record.unserialized ? serializeValue(record.value) : record.value;
					if (value != null) {
						resolvedValue = resolveValue(record.ownerId, record.path, recordValue);
						if (value !== resolvedValue) return;
					}
					return callback(record._id, {
						value: recordValue,
						stamp: record.stamp
					});
				});
				return cursor.closePromised()(Function.prototype);
			}.bind(this));
		}.bind(this));
	}),
	__searchComputed: d(function (keyPath, value, callback) {
		var query;
		if (keyPath) query = { keyPath: keyPath };
		return this.computedDb.invokeAsync('find', query)(function (cursor) {
			return cursor.toArrayPromised()(function (records) {
				records.some(function (record) {
					var recordValue = record.unserialized ? serializeValue(record.value) : record.value;
					if ((value != null) && !filterComputedValue(value, recordValue)) return;
					return callback(record.ownerId + '/' + record.keyPath, {
						value: recordValue,
						stamp: record.stamp
					});
				});
				return cursor.closePromised()(Function.prototype);
			});
		});
	}),

	// Storage import/export
	__exportAll: d(function (destDriver) {
		var count = 0;
		var promise = deferred(
			this.directDb.invokeAsync('find')(function (cursor) {
				return cursor.toArrayPromised()(function (records) {
					return cursor.closePromised()(deferred.map(records, function (record) {
						if (!(++count % 1000)) promise.emit('progress');
						return destDriver._storeRaw('direct', record.ownerId, record.path, {
							value: record.unserialized ? serializeValue(record.value) : record.value,
							stamp: record.stamp
						});
					}, this));
				}.bind(this));
			}.bind(this)),
			this.computedDb.invokeAsync('find')(function (cursor) {
				return cursor.toArrayPromised()(function (records) {
					return cursor.closePromised()(deferred.map(records, function (record) {
						if (!(++count % 1000)) promise.emit('progress');
						return destDriver._storeRaw('computed', record.keyPath, record.ownerId, {
							value: record.unserialized ? serializeValue(record.value) : record.value,
							stamp: record.stamp
						});
					}, this));
				}.bind(this));
			}.bind(this)),
			this.reducedDb.invokeAsync('find')(function (cursor) {
				return cursor.toArrayPromised()(function (records) {
					return cursor.closePromised()(deferred.map(records, function (record) {
						if (!(++count % 1000)) promise.emit('progress');
						return destDriver._storeRaw('reduced', record.ns, record.path, {
							value: record.unserialized ? serializeValue(record.value) : record.value,
							stamp: record.stamp
						});
					}, this));
				}.bind(this));
			}.bind(this))
		);
		return promise;
	}),
	__clear: d(function () {
		return deferred(
			this.hasOwnProperty('directDb') && this.directDb.invokeAsync('drop'),
			this.hasOwnProperty('computedDb') && this.computedDb.invokeAsync('drop'),
			this.hasOwnProperty('reducedDb') && this.reducedDb.invokeAsync('drop')
		);
	}),
	__drop: d(function () { return this.__clear(); }),

	// Connection related
	__close: d(function () { return deferred(undefined); }),

	// Driver specific
	_getDirect_: d(function (ownerId, path) {
		return this.directDb.invokeAsync('find', { _id: ownerId + (path ? ('/' + path) : '') })(
			function (cursor) {
				return cursor.nextPromised()(function (record) {
					return cursor.closePromised()(record ? {
						value: record.unserialized ? serializeValue(record.value) : record.value,
						stamp: record.stamp
					} : getNull);
				}.bind(this));
			}.bind(this)
		);
	}),
	_getComputed_: d(function (ownerId, keyPath) {
		return this.computedDb.invokeAsync('find', { _id: keyPath + ':' + ownerId })(
			function (cursor) {
				return cursor.nextPromised()(function (record) {
					return cursor.closePromised()(record ? {
						value: record.unserialized ? serializeValue(record.value) : record.value,
						stamp: record.stamp
					} : getNull);
				});
			}
		);
	}),
	_getReduced_: d(function (key) {
		return this.reducedDb.invokeAsync('find', { _id: key })(function (cursor) {
			return cursor.nextPromised()(function (record) {
				return cursor.closePromised()(record ? {
					value: record.unserialized ? serializeValue(record.value) : record.value,
					stamp: record.stamp
				} : getNull);
			});
		});
	}),
	_storeDirect_: d(function (ownerId, path, data) {
		var unserialized = Boolean(data.value && isUnserializable(data.value[0]));
		var record = {
			ownerId: ownerId,
			value: unserialized ? unserializeValue(data.value) : data.value,
			stamp: data.stamp
		};
		if (path) {
			record.path = path;
			record.keyPath = resolveKeyPath(ownerId + '/' + path);
		}
		if (unserialized) record.unserialized = true;
		return this.directDb.invokeAsync('update', { _id: ownerId + (path ? ('/' + path) : '') },
			record, updateOpts);
	}),
	_storeComputed_: d(function (ownerId, keyPath, data) {
		var unserialized = Boolean(data.value && isUnserializable(data.value[0]));
		var record = {
			ownerId: ownerId,
			keyPath: keyPath,
			value: unserialized ? unserializeValue(data.value) : data.value,
			stamp: data.stamp
		};
		if (unserialized) record.unserialized = true;
		return this.computedDb.invokeAsync('update', { _id: keyPath + ':' + ownerId },
			record, updateOpts);
	}),
	_storeReduced_: d(function (ns, path, data) {
		var unserialized = Boolean(data.value && isUnserializable(data.value[0]));
		var record = {
			ns: ns,
			value: unserialized ? unserializeValue(data.value) : data.value,
			stamp: data.stamp
		};
		if (path) record.path = path;
		if (unserialized) record.unserialized = true;
		return this.reducedDb.invokeAsync('update', { _id: ns + (path ? ('/' + path) : '') },
			record, updateOpts);
	}),
	_loadDirect_: d(function (query, filter) {
		return this.directDb.invokeAsync('find', query)(function (cursor) {
			return cursor.toArrayPromised()(function (records) {
				var result = create(null);
				records.forEach(function (record) {
					if (filter && !filter(record.ownerId, record.path)) return; // filtered
					result[record._id] = {
						value: record.unserialized ? serializeValue(record.value) : record.value,
						stamp: record.stamp
					};
				});
				return cursor.closePromised()(result);
			}.bind(this));
		}.bind(this));
	})
}, lazy({
	directDb: d(function () {
		return this.driver.mongoDb.invokeAsync('collection', this.name);
	}),
	computedDb: d(function () {
		return this.driver.mongoDb.invokeAsync('collection', this.name + '-computed');
	}),
	reducedDb: d(function () {
		return this.driver.mongoDb.invokeAsync('collection', this.name + '-reduced');
	})
})));
